### PR TITLE
plugin dev  for cloudtrail tags check

### PR DESCRIPTION
**Pull request by abdullahaslam306**

### CODE REVIEW

File: plugins/aws/cloudtrail/cloudtrailHasTags.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'CloudTrail Has Tags',
+    category: 'CloudTrail',
+    domain: 'Compliance',
+    description: 'Ensure that Cloudtrail have tags',
+    more_info: 'Tags help you to group resources together that are related to or associated with each other. It is a best practice to tag cloud resources to better organize and gain visibility into their usage.',
+    recommended_action: 'Modify  cloudtrail and tags',
```
Comment (alphadev4) [inline plugins/aws/cloudtrail/cloudtrailHasTags.js:10]:
```suggestion
    recommended_action: 'Modify  cloudtrail and add tags',
```

File: plugins/aws/cloudtrail/cloudtrailHasTags.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    link: 'https://docs.aws.amazon.com/awscloudtrail/latest/APIReference/API_AddTags.html',
+    apis: ['CloudTrail:describeTrails', 'CloudTrail:listTags'],
+
+    run: function(cache, settings, callback) {
+        var results = [];
+        var source = {};
+        var regions = helpers.regions(settings);
+
+        async.each(regions.cloudtrail, function(region, rcb){
+
```
Comment (alphadev4) [inline plugins/aws/cloudtrail/cloudtrailHasTags.js:20]:
```suggestion
```

File: plugins/aws/cloudtrail/cloudtrailHasTags.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            if (!describeTrails) return rcb();
+
+            if (describeTrails.err || !describeTrails.data) {
+                helpers.addResult(results, 3,
+                    `Unable to query for trails: ${helpers.addError(describeTrails)}`, region);
+                return rcb();
+            }
+
+            if (!describeTrails.data.length) {
+                helpers.addResult(results, 0, 'CloudTrail is not enabled', region);
```
Comment (alphadev4) [inline plugins/aws/cloudtrail/cloudtrailHasTags.js:33]:
```suggestion
                helpers.addResult(results, 0, 'No cloud trails found', region);
```

File: plugins/aws/cloudtrail/cloudtrailHasTags.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+                if (listTags.err || !listTags.data) {
+                    helpers.addResult(results, 3,
+                        `Unable to query for listTags api: ${helpers.addError(listTags)}`, region);
+                    return rcb();
+                }
+
+                if (!listTags.data.ResourceTagList || 
+                !listTags.data.ResourceTagList[0].TagsList || 
+                !listTags.data.ResourceTagList[0].TagsList.length){
+                    helpers.addResult(results, 2, 'Cloudtrail does not have tags associated.', region, trail.TrailARNs);
```
Comment (alphadev4) [inline plugins/aws/cloudtrail/cloudtrailHasTags.js:51]:
```suggestion
                    helpers.addResult(results, 2, 'Cloudtrail does not have tags associated', region, trail.TrailARNs);
```

**Review by alphadev4 — CHANGES_REQUESTED** [review]

File: plugins/aws/cloudtrail/cloudtrailHasTags.js [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+                } else {
+                    helpers.addResult(results, 0, 'Cloudtrail have tags associated.', region, trail.TrailARNs) ;
```
[the remark this answers is not in the document]
Comment (alphadev4) [reply plugins/aws/cloudtrail/cloudtrailHasTags.js:52]: No full stops at the end of results

File: plugins/aws/cloudtrail/cloudtrailHasTags.spec.js [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+            expect(results.length).to.equal(1);
+            expect(results[0].status).to.equal(2);
+            done();
+        });
+    });
+   it('should Passing result if trails have tags', function (done) {
+        const cache = createCache([trails[0]], listTags[1]);
+        cloudtrailHasTags.run(cache, {}, (err, results) => {
+            expect(results.length).to.equal(1);
+            expect(results[0].status).to.equal(0);
```
Comment (alphadev4) [inline plugins/aws/cloudtrail/cloudtrailHasTags.spec.js:120]:
```suggestion
            expect(results[0].status).to.equal(0);
            expect(results[0].message).to.include('Cloudtrail has tags');
```

Comment (alphadev4) [inline plugins/aws/cloudtrail/cloudtrailHasTags.spec.js:120]:
```suggestion
            expect(results[0].status).to.equal(0);
            expect(results[0].message).to.include('Cloudtrail has tags');
```

Comment (alphadev4) [reply plugins/aws/cloudtrail/cloudtrailHasTags.spec.js:120]: Make sure to include message check in all the spec files

File: plugins/aws/cloudtrail/cloudtrailHasTags.js (hunk 5 above)
Comment (alphadev4) [inline plugins/aws/cloudtrail/cloudtrailHasTags.js:52]:
```suggestion
                    helpers.addResult(results, 0, 'Cloudtrail has tags', region, trail.TrailARNs) ;
```

Comment (alphadev4) [reply plugins/aws/cloudtrail/cloudtrailHasTags.js:52]: no full stop at the end of results

File: plugins/aws/cloudtrail/cloudtrailHasTags.js [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+                if (listTags.err || !listTags.data) {
+                    helpers.addResult(results, 3,
+                        `Unable to query for listTags api: ${helpers.addError(listTags)}`, region);
+                    return rcb();
+                }
+
+                if (!listTags.data.ResourceTagList || 
+                !listTags.data.ResourceTagList[0].TagsList || 
+                !listTags.data.ResourceTagList[0].TagsList.length){
+                    helpers.addResult(results, 2, 'Cloudtrail does not have tags associated.', region, trail.TrailARNs);
```
Comment (alphadev4) [inline plugins/aws/cloudtrail/cloudtrailHasTags.js:50]:
```suggestion
                    helpers.addResult(results, 2, 'Cloudtrail does not have tags', region, trail.TrailARNs);
```

**Review by alphadev4 — CHANGES_REQUESTED** [review]

**Review by alphadev4 — APPROVED** [review]

File: plugins/aws/cloudtrail/cloudtrailHasTags.js [hunk 8]
```diff
@@ -0,0 +1,60 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'CloudTrail Has Tags',
+    category: 'CloudTrail',
+    domain: 'Compliance',
+    description: 'Ensure that Cloudtrail have tags',
```
Comment (AkhtarAmir) [inline plugins/aws/cloudtrail/cloudtrailHasTags.js:8]:
```suggestion
    description: 'Ensure that AWS CloudTrail trails have tags associated.',
```

File: plugins/aws/cloudtrail/cloudtrailHasTags.js [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    more_info: 'Tags help you to group resources together that are related to or associated with each other. It is a best practice to tag cloud resources to better organize and gain visibility into their usage.',
+    recommended_action: 'Modify  cloudtrail and add tags',
```
Comment (AkhtarAmir) [inline plugins/aws/cloudtrail/cloudtrailHasTags.js:10]:
```suggestion
    recommended_action: 'Modify CloudTrail trails and add tags.',
```

File: plugins/aws/cloudtrail/cloudtrailHasTags.js [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+        async.each(regions.cloudtrail, function(region, rcb){
+            var describeTrails = helpers.addSource(cache, source,
+                ['cloudtrail', 'describeTrails', region]);
+
+            if (!describeTrails) return rcb();
+
+            if (describeTrails.err || !describeTrails.data) {
+                helpers.addResult(results, 3,
+                    `Unable to query for trails: ${helpers.addError(describeTrails)}`, region);
```
Comment (AkhtarAmir) [inline plugins/aws/cloudtrail/cloudtrailHasTags.js:27]:
```suggestion
                    `Unable to query for CloudTrail trails: ${helpers.addError(describeTrails)}`, region);
```

File: plugins/aws/cloudtrail/cloudtrailHasTags.js [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+                return rcb();
+            }
+
+            if (!describeTrails.data.length) {
+                helpers.addResult(results, 0, 'CloudTrail is not enabled', region);
+                return rcb();
+            }
+            for (let trail of describeTrails.data){
```
Comment (AkhtarAmir) [inline plugins/aws/cloudtrail/cloudtrailHasTags.js:35]:
```suggestion
            }

            for (let trail of describeTrails.data){
```

File: plugins/aws/cloudtrail/cloudtrailHasTags.js [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+                if (!trail.TrailARN) return rcb();
```
Comment (AkhtarAmir) [inline plugins/aws/cloudtrail/cloudtrailHasTags.js:36]:
```suggestion
                if (!trail.TrailARN) continue;
```

File: plugins/aws/cloudtrail/cloudtrailHasTags.js [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+                
+                let listTags = helpers.addSource(cache, source,
+                    ['cloudtrail', 'listTags', region, trail.TrailARN]);
+
+                if (listTags.err || !listTags.data) {
+                    helpers.addResult(results, 3,
+                        `Unable to query for listTags api: ${helpers.addError(listTags)}`, region);
+                    return rcb();
```
Comment (AkhtarAmir) [inline plugins/aws/cloudtrail/cloudtrailHasTags.js:44]:
```suggestion
                    continue;
```

File: plugins/aws/cloudtrail/cloudtrailHasTags.js [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+                }
+
+                if (!listTags.data.ResourceTagList || 
+                !listTags.data.ResourceTagList[0].TagsList || 
+                !listTags.data.ResourceTagList[0].TagsList.length){
+                    helpers.addResult(results, 2, 'Cloudtrail does not have tags', region, trail.TrailARNs);
+                } else {
+                    helpers.addResult(results, 0, 'Cloudtrail has tags', region, trail.TrailARNs) ;
```
Comment (AkhtarAmir) [inline plugins/aws/cloudtrail/cloudtrailHasTags.js:52]:
```suggestion
                    helpers.addResult(results, 0, 'Cloudtrail has tags', region, trail.TrailARNs);
```

File: plugins/aws/cloudtrail/cloudtrailHasTags.js [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+                    ['cloudtrail', 'listTags', region, trail.TrailARN]);
+
+                if (listTags.err || !listTags.data) {
+                    helpers.addResult(results, 3,
+                        `Unable to query for listTags api: ${helpers.addError(listTags)}`, region);
+                    return rcb();
+                }
+
+                if (!listTags.data.ResourceTagList || 
+                !listTags.data.ResourceTagList[0].TagsList ||
```
Comment (AkhtarAmir) [inline plugins/aws/cloudtrail/cloudtrailHasTags.js:48]:
```suggestion
                    !listTags.data.ResourceTagList[0].TagsList || 
```

File: plugins/aws/cloudtrail/cloudtrailHasTags.js [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+
+                if (listTags.err || !listTags.data) {
+                    helpers.addResult(results, 3,
+                        `Unable to query for listTags api: ${helpers.addError(listTags)}`, region);
+                    return rcb();
+                }
+
+                if (!listTags.data.ResourceTagList || 
+                !listTags.data.ResourceTagList[0].TagsList || 
+                !listTags.data.ResourceTagList[0].TagsList.length){
```
Comment (AkhtarAmir) [inline plugins/aws/cloudtrail/cloudtrailHasTags.js:49]:
```suggestion
                    !listTags.data.ResourceTagList[0].TagsList.length){
```

File: plugins/aws/cloudtrail/cloudtrailHasTags.js [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+                if (listTags.err || !listTags.data) {
+                    helpers.addResult(results, 3,
+                        `Unable to query for listTags api: ${helpers.addError(listTags)}`, region);
+                    continue;
+                }
+
+                if (!listTags.data.ResourceTagList || 
+                    !listTags.data.ResourceTagList[0].TagsList || 
+                    !listTags.data.ResourceTagList[0].TagsList.length){
+                    helpers.addResult(results, 2, 'Cloudtrail does not have tags', region, trail.TrailARNs);
```
Comment (AkhtarAmir) [inline plugins/aws/cloudtrail/cloudtrailHasTags.js:51]:
```suggestion
                    helpers.addResult(results, 2, 'CloudTrail trail does not have tags', region, trail.TrailARN);
```

File: plugins/aws/cloudtrail/cloudtrailHasTags.js [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+                    ['cloudtrail', 'listTags', region, trail.TrailARN]);
+
+                if (listTags.err || !listTags.data) {
+                    helpers.addResult(results, 3,
+                        `Unable to query for listTags api: ${helpers.addError(listTags)}`, region);
+                    continue;
+                }
+
+                if (!listTags.data.ResourceTagList || 
+                    !listTags.data.ResourceTagList[0].TagsList ||
```
Comment (AkhtarAmir) [inline plugins/aws/cloudtrail/cloudtrailHasTags.js:49]:
```suggestion
                if (!listTags.data.ResourceTagList[0].TagsList || 
```

File: plugins/aws/cloudtrail/cloudtrailHasTags.js [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+                return rcb();
+            }
+
+            for (let trail of describeTrails.data){
+                if (!trail.TrailARN) continue;
+                
+                let listTags = helpers.addSource(cache, source,
+                    ['cloudtrail', 'listTags', region, trail.TrailARN]);
+
+                if (listTags.err || !listTags.data) {
```
Comment (AkhtarAmir) [inline plugins/aws/cloudtrail/cloudtrailHasTags.js:42]:
```suggestion
                if (listTags.err || !listTags.data || !listTags.data.ResourceTagList || !listTags.data.ResourceTagList.length) {
```

File: plugins/aws/cloudtrail/cloudtrailHasTags.js [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+                    helpers.addResult(results, 3,
+                        `Unable to query for listTags api: ${helpers.addError(listTags)}`, region);
+                    continue;
+                }
+
+                if (!listTags.data.ResourceTagList[0].TagsList || 
+                    !listTags.data.ResourceTagList[0].TagsList.length){
+                    helpers.addResult(results, 2, 'CloudTrail trail does not have tags', region, trail.TrailARN);
+                } else {
+                    helpers.addResult(results, 0, 'Cloudtrail has tags', region, trail.TrailARNs);
```
Comment (AkhtarAmir) [inline plugins/aws/cloudtrail/cloudtrailHasTags.js:52]:
```suggestion
                    helpers.addResult(results, 0, 'CloudTrail trail has tags', region, trail.TrailARN);
```

File: plugins/aws/cloudtrail/cloudtrailHasTags.spec.js [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+describe('cloudtrailHasTags', function () {
+    describe('run', function () {
+
+        it('should UNKNOWN if unable to query for trails', function (done) {
+            const cache = createErrorCache();
+            cloudtrailHasTags.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(3);
+                expect(results[0].region).to.equal('us-east-1');
+                expect(results[0].message).to.include('Unable to query for trails');
```
Comment (AkhtarAmir) [inline plugins/aws/cloudtrail/cloudtrailHasTags.spec.js:88]:
```suggestion
                expect(results[0].message).to.include('Unable to query for CloudTrail trails');
```

File: plugins/aws/cloudtrail/cloudtrailHasTags.js [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+
+            for (let trail of describeTrails.data){
+                if (!trail.TrailARN) continue;
+                
+                let listTags = helpers.addSource(cache, source,
+                    ['cloudtrail', 'listTags', region, trail.TrailARN]);
+
+                if (listTags.err || !listTags.data || !listTags.data.ResourceTagList || !listTags.data.ResourceTagList.length) {
+                    helpers.addResult(results, 3,
+                        `Unable to query for listTags api: ${helpers.addError(listTags)}`, region);
```
Comment (AkhtarAmir) [inline plugins/aws/cloudtrail/cloudtrailHasTags.js:44]:
```suggestion
                        `Unable to list trail tags: ${helpers.addError(listTags)}`, region);
```

File: plugins/aws/cloudtrail/cloudtrailHasTags.spec.js [hunk 23]
```diff
@@ -0,0 +1,135 @@
...
+        });
+    });
+
+    it('should Unknown result if unable to query listTags', function (done) {
+        const cache = createCache([trails[0]], null);
+        cloudtrailHasTags.run(cache, {}, (err, results) => {
+            expect(results.length).to.equal(1);
+            expect(results[0].status).to.equal(3);
+            expect(results[0].region).to.equal('us-east-1');
+            expect(results[0].message).to.include('Unable to query for listTags api');
```
Comment (AkhtarAmir) [inline plugins/aws/cloudtrail/cloudtrailHasTags.spec.js:109]:
```suggestion
            expect(results[0].message).to.include('Unable to list trail tags');
```

File: plugins/aws/cloudtrail/cloudtrailHasTags.spec.js [hunk 24]
```diff
@@ -0,0 +1,135 @@
...
+        });
+    });
+
+     it('should Failing result if trails have no tags', function (done) {
+        const cache = createCache([trails[0]], listTags[0]);
+        cloudtrailHasTags.run(cache, {}, (err, results) => {
+            expect(results.length).to.equal(1);
+            expect(results[0].status).to.equal(2);
+            expect(results[0].region).to.equal('us-east-1');
+            expect(results[0].message).to.include('Cloudtrail does not have tags');
```
Comment (AkhtarAmir) [inline plugins/aws/cloudtrail/cloudtrailHasTags.spec.js:120]:
```suggestion
            expect(results[0].message).to.include('CloudTrail trail does not have tags');
```

File: plugins/aws/cloudtrail/cloudtrailHasTags.spec.js [hunk 25]
```diff
@@ -0,0 +1,135 @@
...
+            done();
+        });
+    });
+   it('should Passing result if trails have tags', function (done) {
+        const cache = createCache([trails[0]], listTags[1]);
+        cloudtrailHasTags.run(cache, {}, (err, results) => {
+            expect(results.length).to.equal(1);
+            expect(results[0].status).to.equal(0);
+            expect(results[0].region).to.equal('us-east-1');
+            expect(results[0].message).to.include('Cloudtrail has tags');
```
Comment (AkhtarAmir) [inline plugins/aws/cloudtrail/cloudtrailHasTags.spec.js:130]:
```suggestion
            expect(results[0].message).to.include('CloudTrail trail has tags');
```

**Pull request merged**